consensus/tendermint/tendermint_test.go

```diff
@@ -0,0 +1,277 @@
+		listeners, broadcasters := testListenersAndBroadcasters()
+		app, chain, vals := newApp(), newChain(), newVals()
+
+		vals.addValidator(*val2)
+		vals.addValidator(*val3)
+		vals.addValidator(*val4)
+		vals.addValidator(*nodeAddr)
+
+		algo := New[value, felt.Felt, felt.Felt](*nodeAddr, app, chain, vals, listeners, broadcasters, tm, tm, tm)
+
+		algo.Start()
+		algo.Stop()
+
+		assert.Equal(t, 1, len(algo.scheduledTms))
+		scheduledTm := algo.scheduledTms[0]
+
+		assert.Equal(t, propose, scheduledTm.s)
+		assert.Equal(t, height(0), scheduledTm.h)
+		assert.Equal(t, round(0), scheduledTm.r)
+
+		assert.Equal(t, propose, algo.state.s)
+		assert.Equal(t, height(0), algo.state.h)
+		assert.Equal(t, round(0), algo.state.r)
+	})
+
+	t.Run("OnTimeoutPropose: round zero the node is not the proposer thus send a prevote nil", func(t *testing.T) {
+		listeners, broadcasters := testListenersAndBroadcasters()
+		app, chain, vals := newApp(), newChain(), newVals()
+		// The algo needs to run for a minimum amount of time but it cannot be long enough the state to change
+		// multiple times. This can happen if the timeouts are too small.
+		tm := func(r round) time.Duration {
+			if r == 0 {
+				return time.Nanosecond
+			}
+			return time.Second
+		}
+
+		vals.addValidator(*val2)
+		vals.addValidator(*val3)
+		vals.addValidator(*val4)
+		vals.addValidator(*nodeAddr)
+
+		algo := New[value, felt.Felt, felt.Felt](*nodeAddr, app, chain, vals, listeners, broadcasters, tm, tm, tm)
+
+		expectedHeight, expectedRound := height(0), round(0)
+		expectedPrevoteMsg := Prevote[felt.Felt, felt.Felt]{
+			H:      0,
+			R:      0,
+			ID:     nil,
+			Sender: *nodeAddr,
+		}
+
+		prevoteBroadcaster := broadcasters.PrevoteBroadcaster.(*senderAndReceiver[Prevote[felt.Felt, felt.Felt], value,
+			felt.Felt, felt.Felt])
+
+		algo.Start()
+		time.Sleep(time.Millisecond)
+		algo.Stop()
+
+		prevoteMsg := <-prevoteBroadcaster.mCh
+
+		assert.Equal(t, expectedPrevoteMsg, prevoteMsg)
+		assert.Equal(t, 1, len(algo.messages.prevotes[expectedHeight][expectedRound][*nodeAddr]))
+		assert.Equal(t, expectedPrevoteMsg, algo.messages.prevotes[expectedHeight][expectedRound][*nodeAddr][0])
+
+		assert.Equal(t, prevote, algo.state.s)
+		assert.Equal(t, expectedHeight, algo.state.h)
+		assert.Equal(t, expectedRound, algo.state.r)
+	})
+}
+
+func TestThresholds(t *testing.T) {
+	tests := []struct {
+		n uint
+		q uint
+		f uint
+	}{
+		{1, 1, 0},
+		{2, 2, 0},
+		{3, 2, 0},
+		{4, 3, 1},
+		{5, 4, 1},
+		{6, 4, 1},
+		{7, 5, 2},
+		{11, 8, 3},
+		{15, 10, 4},
+		{20, 14, 6},
+		{100, 67, 33},
+		{150, 100, 49},
+		{2000, 1334, 666},
+		{2509, 1673, 836},
+		{3045, 2030, 1014},
+		{7689, 5126, 2562},
+		{10032, 6688, 3343},
+		{12932, 8622, 4310},
+		{15982, 10655, 5327},
+		{301234, 200823, 100411},
+		{301235, 200824, 100411},
+		{301236, 200824, 100411},
+	}
+
+	for _, test := range tests {
+		assert.Equal(t, test.q, q(test.n))
+		assert.Equal(t, test.f, f(test.n))
+	}
```

Comment:
I think what you'd want to assert here for q function is
- For correctness: 2 * q(test.n) > test.n + f(test.n)
- For efficiency: 2 * (q(test.n) - 1) <= test.n + f(test.n)

Reply:
If i have already manually calculated the q_min and max_f, I don't see the reason to add the equations.